File: python/sorting/quickSort.py
def shuffle(arr, start, end):

    i = start-1
    j = start

    pivot = end

    while (j < end):

        if (arr[j] < arr[pivot]):

            i += 1
            arr[i], arr[j] = arr[j], arr[i]
        j += 1

    arr[i + 1], arr[pivot] = arr[pivot], arr[i + 1]

    return i + 1


def quickSort(arr, start, end):

    if (start < end):
        pivotEle = shuffle(arr, start, end)
        quickSort(arr, start, pivotEle-1)
        quickSort(arr, pivotEle + 1, end)


arr = [2, 4, 5, 6, 1, 3, 4, 8]
start = 0
end = len(arr)-1
print(quickSort(arr, start, end))
print(arr)

File: javascript/sorting/quickSort.js
const solve = (arr) => {
  const swap = (arr, l, r) => {
    [arr[l], arr[r]] = [arr[r], arr[l]];
  };
  const rearrange = (arr, s, e) => {
    let l = s + 1;
    let r = e;
    let pivot = s;
    while (l <= r) {
      if (arr[pivot] >= arr[l]) {
        l++;
      } else if (arr[pivot] < arr[r]) {
        r--;
      } else {
        swap(arr, l, r);
      }
    }
    swap(arr, pivot, r);

    return r;
  };

  const quickSort = (arr, s, e) => {
    if (s == e || s > e) {
      return;
    }

    p = rearrange(arr, s, e);

    quickSort(arr, s, p - 1);
    quickSort(arr, p + 1, e);
  };
  let s = 0;
  let e = arr.length - 1;
  quickSort(arr, s, e);
  return arr;
};

console.log(solve([5, 2, 0, 8, 3]));
